File: DictationAssistant/ImprovedSpeakEngine.vb
Imports System.Globalization
Imports System.IO

Public Class ImprovedSpeakEngine
    Implements ISpeakEngine

    Private Shared ReadOnly AudioFileExtensions As String() = "wav;flac;ape;m4a;opus;acc;mp3;mp2;mp1;ogg;wma;aif;mp4".Split(";"c)

    Private EngineForNoFile As ISpeakEngine
    Private ResForImproving As String
    Public Sub New(EngineForNoFile As ISpeakEngine, ResForImproving As String)
        Me.EngineForNoFile = EngineForNoFile
        Me.ResForImproving = ResForImproving
    End Sub

    Public ReadOnly Property Name As String Implements ISpeakEngine.Name
        Get
            Return "[Improved] " & EngineForNoFile.Name
        End Get
    End Property
    Public ReadOnly Property Culture As CultureInfo Implements ISpeakEngine.Culture
        Get
            Return EngineForNoFile.Culture
        End Get
    End Property


    Private Function FindFile(text As String) As String
        Dim filePath As String
        filePath = text
        filePath = filePath.Replace("\", "")
        filePath = filePath.Replace("/", "")
        filePath = filePath.Replace(":", "")
        filePath = filePath.Replace("*", "")
        filePath = filePath.Replace("?", "")
        filePath = filePath.Replace(Chr(34), "")
        filePath = filePath.Replace("<", "")
        filePath = filePath.Replace(">", "")
        filePath = filePath.Replace("|", "")
        filePath = Path.Combine(ResForImproving, filePath & ".")
        For Each extension In AudioFileExtensions
            If File.Exists(filePath & extension) Then
                Return filePath & extension
            End If
        Next
        Throw New FileNotFoundException($"未找到词组{text}对应的音频文件")
    End Function
    Public Function Speak(Text As String, Param As SpeakParam) As PcmStreamWithInfo Implements ISpeakEngine.Speak
        Try
            Return AudioFileDecodeStream.Create(FindFile(Text))
        Catch ex As Exception
            If EngineForNoFile Is Nothing Then
                Return PcmStreamWithInfo.Empty
            End If
            Return EngineForNoFile.Speak(Text, Param)
        End Try
    End Function
End Class

File: DictationAssistant/ImprovedSpeakEngine.cs
using Microsoft.VisualBasic;
using System;
using System.Globalization;
using System.IO;

namespace DictationAssistant
{
    public class ImprovedSpeakEngine : ISpeakEngine
    {
        private static readonly string[] AudioFileExtensions = "wav;flac;ape;m4a;opus;acc;mp3;mp2;mp1;ogg;wma;aif;mp4".Split(';');

        private ISpeakEngine EngineForNoFile;
        private string ResForImproving;
        public ImprovedSpeakEngine(ISpeakEngine engineForNoFile, string resForImproving)
        {
            this.EngineForNoFile = engineForNoFile;
            this.ResForImproving = resForImproving;
        }

        public string Name
        {
            get
            {
                return "[Improved] " + EngineForNoFile.Name;
            }
        }
        public CultureInfo Culture
        {
            get
            {
                return EngineForNoFile.Culture;
            }
        }


        private string FindFile(string text)
        {
            string filePath;
            filePath = text;
            filePath = filePath.Replace(@"\", "");
            filePath = filePath.Replace("/", "");
            filePath = filePath.Replace(":", "");
            filePath = filePath.Replace("*", "");
            filePath = filePath.Replace("?", "");
            filePath = filePath.Replace("\"", "");
            filePath = filePath.Replace("<", "");
            filePath = filePath.Replace(">", "");
            filePath = filePath.Replace("|", "");
            filePath = Path.Combine(ResForImproving, filePath + ".");
            foreach (var extension in AudioFileExtensions)
            {
                if (File.Exists(filePath + extension))
                    return filePath + extension;
            }
            throw new FileNotFoundException($"未找到词组{text}对应的音频文件");
        }
        public PcmStreamWithInfo Speak(string text, SpeakParam param)
        {
            try
            {
                return AudioFileDecodeStream.Create(FindFile(text));
            }
            catch (Exception)
            {
                if (EngineForNoFile == null)
                    return PcmStreamWithInfo.Empty;
                return EngineForNoFile.Speak(text, param);
            }
        }
    }
}
